File: Leetcode/PermutationinString.cpp
//
// Created by Quyen Linh TA on 30/01/2023.
//

class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();
        if (n > m) return false;
        vector<int> cnt1(26, 0);
        vector<int> cnt2(26, 0);
        for (int i = 0; i < n; i++) {
            cnt1[s1[i] - 'a']++;
            cnt2[s2[i] - 'a']++;
        }
        for (int i = n; i < m; i++) {
            if (cnt1 == cnt2) return true;
            cnt2[s2[i] - 'a']++;
            cnt2[s2[i - n] - 'a']--;
        }
        return cnt1 == cnt2;

    }
};

File: Leetcode/PermutationinString.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        s1_dict = {}
        for i in s1:
            if i not in s1_dict:
                s1_dict[i] = 1
            else:
                s1_dict[i] += 1
        s2_dict = {}
        for i in s2[:len(s1)]:
            if i not in s2_dict:
                s2_dict[i] = 1
            else:
                s2_dict[i] += 1
        if s1_dict == s2_dict:
            return True
        for i in range(len(s1), len(s2)):
            s2_dict[s2[i - len(s1)]] -= 1
            if s2_dict[s2[i - len(s1)]] == 0:
                del s2_dict[s2[i - len(s1)]]
            if s2[i] not in s2_dict:
                s2_dict[s2[i]] = 1
            else:
                s2_dict[s2[i]] += 1
            if s1_dict == s2_dict:
                return True
        return False